### ISSUE TITLE
API request to move user to different organization

### ISSUE THREAD

**Issue by peintnermax** [zitadel/zitadel#8835]
### Preflight Checklist

- [x] I could not find a solution in the existing issues, docs, nor discussions
- [x] I have joined the [ZITADEL chat](https://zitadel.com/chat)

### Describe your problem

A user needs the ability to seamlessly switch between organizations, keeping its original propertiers and settings.

### Describe your ideal solution

An API request to be able to set a new organization as owner of the user resource.

### Environment

ZITADEL Cloud

**Comment by dhrp** [comment]
Just as an addition: I'm also looking for this feature; in a use-case that I think is also common:

**Allowing user self-registration; and then converting that user to an organization.**

Not unlike Github, you may start as an individual user, but later realize you need others to join you, and you need to establish an organization relationship with the provider. i.e. invite other users and so forth. Currently the best? option is to create an organization from the start, but it creates a lot of overhead possibly clutter from one-time users. Moving the user into a new organization would help solve this problem.

**Comment by d-sharpe** [comment]
> Just as an addition: I'm also looking for this feature; in a use-case that I think is also common:
> 
> **Allowing user self-registration; and then converting that user to an organization.**
> 
> Not unlike Github, you may start as an individual user, but later realize you need others to join you, and you need to establish an organization relationship with the provider. i.e. invite other users and so forth. Currently the best? option is to create an organization from the start, but it creates a lot of overhead possibly clutter from one-time users. Moving the user into a new organization would help solve this problem.

Would it be fair to assume that the current flow to accomplish this would be create new organisation for user(s) --> export data from zitadel using "admin/v1/export" endpoint --> extract the humanUser(s) of interest --> delete the user(s) via "v2/users/:userId" --> import/add new human user via "v2/users/human" with extracted user data amended with new organisation details?

**Comment by hifabienne** [comment]
> > Just as an addition: I'm also looking for this feature; in a use-case that I think is also common:
> > **Allowing user self-registration; and then converting that user to an organization.**
> > Not unlike Github, you may start as an individual user, but later realize you need others to join you, and you need to establish an organization relationship with the provider. i.e. invite other users and so forth. Currently the best? option is to create an organization from the start, but it creates a lot of overhead possibly clutter from one-time users. Moving the user into a new organization would help solve this problem.
> 
> Would it be fair to assume that the current flow to accomplish this would be create new organisation for user(s) --> export data from zitadel using "admin/v1/export" endpoint --> extract the humanUser(s) of interest --> delete the user(s) via "v2/users/:userId" --> import/add new human user via "v2/users/human" with extracted user data amended with new organisation details?

Yes I think the general flow would be sufficient at the moment. Disclaimer: I have not tested it myself.

**Comment by ArnauAregall** [comment]
> Would it be fair to assume that the current flow to accomplish this would be create new organisation for user(s) --> export data from zitadel using "admin/v1/export" endpoint --> extract the humanUser(s) of interest --> delete the user(s) via "v2/users/:userId" --> import/add new human user via "v2/users/human" with extracted user data amended with new organisation details?

Hello @d-sharpe, did you test if this strategy worked out? Similar needs here. Thanks!

**Comment by JakobStadlhuber** [comment]
we also looking for a solution to switch a user via the api to a different org

**Comment by micael-pn** [comment]
Hello,

We are also looking for this feature. Currently, we use the following strategy : 

1. Fetch the user information using `/v2/users/:userId` GET route
2. Search for the last `user.human.password.changed` event using `/admin/v1/events/_search` GET route in order to retrieve the `encodedHash` password
3. Delete the user from the original organization using `/v2/users/:userId` DELETE route
4. Create the user in the new organization using `v2/users/new` POST

It would be more convenient and secure to use a dedicated endpoint.

**Comment by elmawardy** [comment]
> Just as an addition: I'm also looking for this feature; in a use-case that I think is also common:
> 
> **Allowing user self-registration; and then converting that user to an organization.**
> 
> Not unlike Github, you may start as an individual user, but later realize you need others to join you, and you need to establish an organization relationship with the provider. i.e. invite other users and so forth. Currently the best? option is to create an organization from the start, but it creates a lot of overhead possibly clutter from one-time users. Moving the user into a new organization would help solve this problem.

I've tried to accomplish that by exporting the user > delete the user > recreate the user as mentioned by @d-sharpe , but the registration process was disrupted showing something like "User not found" error. instead the following flow worked:
- Keep the user on the main Zitadel org
- Create a project inside the new org or grant existing project with some role ie "x"
- Attach that role to the new user.
- Add an action so that you know the "Authorizations" of the user and get which org does the user has the "x" role attached.

following are mentions of some API calls used for the process:

Create org:
POST /v2/organizations

Create project inside org:
POST /zitadel.project.v2.ProjectService/CreateProject
============= OR ==============
Grant project to the new org
POST /management/v1/projects/<project id>/grants

Add authorization to project for user:
POST /zitadel.authorization.v2.AuthorizationService/CreateAuthorization



then inside the v1 action:
POST /zitadel.authorization.v2.AuthorizationService/ListAuthorizations

**Comment by FredrikPAC** [comment]
Would like to +1 this.
It becomes a larger problem when relying on the event callback from zitadel user deletion to scrub user pii data.
In essence the deletion event maps poorly, since deleting wasn't the intent. It was moving/migrating.
